### PR TITLE
 Delay restart after installing/uninstalling/updating plugins

### LINKED ISSUE #2419: BUG: Downloading plugins makes error when original zip file existing
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I am using the latest version of Flow Launcher.

### Problem Description

I want to install a plugin named GamesLauncher, but in some reason only zip file without install. When I try to install again, it reports error of download. The log file says:
`ERROR - PluginsManager.InstallOrUpdate - An error occurred while downloading plugin
EXCEPTION OCCURS: System.IO.IOException: The file 'C:\Users\(MyUser)\AppData\Local\FlowLauncher\app-1.16.2\UserData\Plugins\GamesLauncher-1.6.0.zip' already exists.`
Seems that Flow.launcher has no ways to handle an existing zip file. It should be reused or redownload after deleting.

### To Reproduce

It may hard to reproduce... I have remember that I download 2 plugins also making errors, but it because of my bad networks. Regulating my networks, I successfully downloaded the other plugin, but GamesLauncher not and cause errors below. I guess after GamesLauncher.zip download and before GamesLauncher install Flow.Launcher reboot because of the other plugin installation.

### Screenshots

_No response_

### Flow Launcher Version

1.16.2

### Windows Build Number

10.0.22000.708

### Error Log

<details>

```
ERROR - PluginsManager.InstallOrUpdate - An error occurred while downloading plugin
EXCEPTION OCCURS: System.IO.IOException: The file 'C:\Users\(MyUser)\AppData\Local\FlowLauncher\app-1.16.2\UserData\Plugins\GamesLauncher-1.6.0.zip' already exists.

```
</details>

### PR DESCRIPTION
**Pull request by VictoriousRaptor**
## Feature

- Delay restart after installing/uninstalling/updating plugins.

### Core
-  **Move Install/Uninstall plugin logic to `Core.PluginManager`**
- Disallow modifying a plugin twice in a session.

### Plugins Manger
- New option for Plugins Manger: `AutoRestartAfterChanging`. Default is `false`. (Should we set to true to keep current behavior?)
- Exclude installed plugins from `pm install`
- Exclude modified(i.e. updated/installed/uninstalled in this session) plugins from `pm install` and `pm update` results
- Show plugin remote icons in `pm Install` results.
- Add glyphs for pm context menu
- Reduce messageboxes to improve consistency
- Fix #2419 

## Tests:
- [x] Different message boxes and notification are shown when `AutoRestartAfterChanging == false`.
- [x] plugins can be installed/uninstalled after restart
- [x] Uninstall a plugin, then install again before restarting not allowed.
- [x] Installing twice before restarting not allowed. 
- [x] Existing zip deleted when downloading zip.

### CODE REVIEW

File: Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs [hunk 1]
```diff
@@ -136,13 +146,11 @@ internal async Task InstallOrUpdateAsync(UserPlugin plugin)
             {
                 await Http.DownloadAsync(plugin.UrlDownload, filePath).ConfigureAwait(false);
 
-                Context.API.ShowMsg(Context.API.GetTranslation("plugin_pluginsmanager_downloading_plugin"),
-                     string.Format(Context.API.GetTranslation("plugin_pluginsmanager_download_success"), plugin.Name));
-
```
Comment (VictoriousRaptor) [inline Plugins/Flow.Launcher.Plugin.PluginsManager/PluginsManager.cs:141]: Download success notification removed.

**Comment by VictoriousRaptor** [comment]
Oh no looks like spell check is really broken...

**Comment by jjw24** [comment]
We will need to handle another obscure scenario where user chooses to do manual restart, install a plugin and then for some reason uninstalls the same plugin, this will freak flow out a little. Same thing but for vice versa will also happen- uninstall and install the same plugin.

File: Plugins/Flow.Launcher.Plugin.PluginsManager/Languages/en.xaml [hunk 2]
```diff
@@ -21,15 +23,19 @@
...
     <system:String x:Key="plugin_pluginsmanager_update_failed_title">Plugin Manifest Download Failed</system:String>
     <system:String x:Key="plugin_pluginsmanager_update_failed_subtitle">Please check if you can connect to github.com. This error means you may not be able to install or update plugins.</system:String>
+    <system:String x:Key="plugin_pluginsmanager_update_success_restart">Plugin {0} successfully updated. Restarting Flow, please wait...</system:String>
     <system:String x:Key="plugin_pluginsmanager_install_unknown_source_warning_title">Installing from an unknown source</system:String>
     <system:String x:Key="plugin_pluginsmanager_install_unknown_source_warning">You are installing this plugin from an unknown source and it may contain potential risks!{0}{0}Please ensure you understand where this plugin is from and that it is safe.{0}{0}Would you like to continue still?{0}{0}(You can switch off this warning via settings)</system:String>
-
-    <!--  Controls  -->
+    
+    <system:String x:Key="plugin_pluginsmanager_install_success_no_restart">Plugin {0} successfully installed. Please manually restart Flow.</system:String>
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.PluginsManager/Languages/en.xaml:36]:
```suggestion
    <system:String x:Key="plugin_pluginsmanager_install_success_no_restart">Plugin {0} successfully installed. Please restart Flow.</system:String>
```

File: Plugins/Flow.Launcher.Plugin.PluginsManager/Languages/en.xaml [hunk 3]
```diff
@@ -21,15 +23,19 @@
...
+    <system:String x:Key="plugin_pluginsmanager_uninstall_success_no_restart">Plugin {0} successfully uninstalled. Please manually restart Flow.</system:String>
+    <system:String x:Key="plugin_pluginsmanager_update_success_no_restart">Plugin {0} successfully updated. Please manually restart Flow.</system:String>
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.PluginsManager/Languages/en.xaml:38]:
```suggestion
    <system:String x:Key="plugin_pluginsmanager_update_success_no_restart">Plugin {0} successfully updated. Please restart Flow.</system:String>
```

File: Plugins/Flow.Launcher.Plugin.PluginsManager/Languages/en.xaml [hunk 4]
```diff
@@ -21,15 +23,19 @@
...
+    <system:String x:Key="plugin_pluginsmanager_update_success_restart">Plugin {0} successfully updated. Restarting Flow, please wait...</system:String>
     <system:String x:Key="plugin_pluginsmanager_install_unknown_source_warning_title">Installing from an unknown source</system:String>
     <system:String x:Key="plugin_pluginsmanager_install_unknown_source_warning">You are installing this plugin from an unknown source and it may contain potential risks!{0}{0}Please ensure you understand where this plugin is from and that it is safe.{0}{0}Would you like to continue still?{0}{0}(You can switch off this warning via settings)</system:String>
-
-    <!--  Controls  -->
+    
+    <system:String x:Key="plugin_pluginsmanager_install_success_no_restart">Plugin {0} successfully installed. Please manually restart Flow.</system:String>
+    <system:String x:Key="plugin_pluginsmanager_uninstall_success_no_restart">Plugin {0} successfully uninstalled. Please manually restart Flow.</system:String>
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.PluginsManager/Languages/en.xaml:37]:
```suggestion
    <system:String x:Key="plugin_pluginsmanager_uninstall_success_no_restart">Plugin {0} successfully uninstalled. Please restart Flow.</system:String>
```

File: Flow.Launcher.Plugin/Interfaces/IPublicAPI.cs [hunk 5]
```diff
@@ -107,15 +107,15 @@ public interface IPublicAPI
         /// </summary>
         /// <param name="title">Message title</param>
         /// <param name="subTitle">Message subtitle</param>
-        /// <param name="iconPath">Message icon path (relative path to your plugin folder)</param>
+        /// <param name="iconPath">Full path to icon</param>
```
Comment (jjw24) [inline Flow.Launcher.Plugin/Interfaces/IPublicAPI.cs:110]: Is this comment change correct? Is it relative or full path? I thought it's still some like `./img/icon.png`?

Comment (VictoriousRaptor) [reply Flow.Launcher.Plugin/Interfaces/IPublicAPI.cs:110]: It's the full path.

**Comment by jjw24** [comment]
> We will need to handle an obscure scenario where user chooses to do manual restart, install a plugin and then for some reason uninstalls the same plugin, this will freak flow out a little. Same thing but for vice versa will also happen- uninstall and install the same plugin.

Best I can think of that is easy to solve is persisting three type of lists installed/uninstalled/updated in memory and each action should check against this list first.

File: Flow.Launcher.Plugin/Interfaces/IPublicAPI.cs (hunk 5 above)
Comment (taooceros) [reply Flow.Launcher.Plugin/Interfaces/IPublicAPI.cs:110]: relative path also works

Comment (VictoriousRaptor) [reply Flow.Launcher.Plugin/Interfaces/IPublicAPI.cs:110]: > relative path also works

Are you sure? iirc I changed this line cuz I found a Python plugin had no icon. maybe it's a Python side issue?

Comment (taooceros) [reply Flow.Launcher.Plugin/Interfaces/IPublicAPI.cs:110]: https://github.com/Flow-Launcher/Flow.Launcher/blob/026e5f5b0967113ac92c00ad92fa031454834295/Flow.Launcher.Plugin/Result.cs#L63-L81

Maybe the json parser doesn't abide this setter?

**Comment by VictoriousRaptor** [comment]
> > We will need to handle an obscure scenario where user chooses to do manual restart, install a plugin and then for some reason uninstalls the same plugin, this will freak flow out a little. Same thing but for vice versa will also happen- uninstall and install the same plugin.
> 
> Best I can think of that is easy to solve is persisting three type of lists installed/uninstalled/updated in memory and each action should check against this list first.

You can't uninstall right after installing since the installed plugin is not added to list yet. Installing after uninstalling throws exception though.

**Comment by jjw24** [comment]
>You can't uninstall right after installing since the installed plugin is not added to list yet. Installing after uninstalling throws exception though

Sorry I meant installing, then installing the same plugin again without restarting.

**Comment by jjw24** [comment]
hey @VictoriousRaptor looks like you added merging the plugins manager functionality  to this PR, is it possible to branch this change out so this PR is just for the delay restart ?

**Review by jjw24 — APPROVED** [review]
👍

**Pull request merged**